bqskit/ir/circuit.py

```diff
@@ -2143,193 +2221,145 @@ def surround(
-                    if bounding_region is not None:
-                        if (cycle_index, qudit_index) not in bounding_region:
+            for point in self.next(node).union(self.prev(node)):
+                # Create new region by adding the gate at this point
+                region_bldr = {k: v for k, v in node.items()}
+                op = self[point]
+                valid_region = True
+                for qudit in op.location:
+                    if qudit not in region_bldr:
+                        region_bldr[qudit] = CycleInterval(point[0], point[0])
+
+                    elif point[0] < region_bldr[qudit][0]:
+                        # Check for gates in the middle not in region
+                        if any(
+                            not self.is_point_idle((i, qudit))
+                            for i in range(point[0] + 1, region_bldr[qudit][0])
+                        ):
+                            valid_region = False
                             break
 
-                    # Stop when exploring previously explored points
-                    point = CircuitPoint(cycle_index, qudit_index)
-                    if point in node[3]:
-                        break
-                    node[3].add(point)
-
-                    # Continue until next operation
-                    if self.is_point_idle(point):
-                        continue
-                    op: Operation = self[cycle_index, qudit_index]
+                        # Absorb Single-qudit gates
+                        index = point[0]
+                        while index > 0:
+                            if not self.is_point_idle((index - 1, qudit)):
+                                prev_op = self[index - 1, qudit]
+                                if len(prev_op.location) != 1:
+                                    break
+                            index -= 1
+
+                        region_bldr[qudit] = CycleInterval(
+                            index,
+                            region_bldr[qudit][1],
+                        )
 
-                    # Gates already in region stop the half_wire
-                    if (cycle_index, op) in node[1]:
-                        break
+                    elif point[0] > region_bldr[qudit][1]:
+                        # Check for gates in the middle not in region
+                        if any(
+                            not self.is_point_idle((i, qudit))
+                            for i in range(region_bldr[qudit][1] + 1, point[0])
+                        ):
+                            valid_region = False
+                            break
 
-                    # Gates already accounted for stop the half_wire
-                    if (cycle_index, op) in absorbed_gates:
-                        break
+                        # Absorb Single-qudit gates
+                        index = point[0]
+                        while index < self.num_cycles - 1:
```

Comment:
Maybe the parts that do "Absorb Single-qudit gates" and "Check for gates in the middle not in region" should be sub-functions on there own.

Reply:
Yeah, using the flag "valid_region" gave me that same thought, but I can't see a way that makes it any simpler. How are you thinking to do this?

Reply:
It might actually be most clear as is, since the searches over cycles go in opposite directions. Does the `valid_region = False` on line 2327 ever run in practice? Otherwise LGTM

Reply:
Yeah, that flag is incredibly important. It prevents us from needing to do full, expensive checks all the time. Think of a diamond of cnots (4 qubits, 1->2 ; 0->1, 2->3 ; 1->2). The search can find a U-shape region that goes around one of the cnots in the middle layer. We need to mark this as invalid and prune. If that shape belongs to the largest (goal) region, then it's fine. It's provable that any valid region has a path in search space of smaller valid regions leading up to it.